### PR TITLE
Lookup type 6

### PR DESCRIPTION
**Pull request by RoelN**
This PR adds lookup type 6. This is a way for fonts to look before and after the character as it were. For instance, the `frac` layout feature is usually a type 6 lookup that basically works like this:

> If there's a **slash**, see if there's a **number** before it, see if there's a **number** after it. If so, exchange these characters by these glyphs.

In the case of a fraction like `1/2`, the 1 would be changed into superscript 1, the slash optionally for a different designed slash, and the 2 for a subscript 2, making a ½

In this example, the slash is the `input`, the `1` is the backtrack, and the `2` is the lookahead.

This PR exposes the input, backtrack and lookhead so an interface can figure out how to display it.

### CODE REVIEW

File: src/fondue/Fondue.js [hunk 1]
```diff
@@ -712,14 +713,30 @@ export default class Fondue {
...
+									let g = range.startGlyphID;
+									g < range.endGlyphID + 1;
+									g++
+								) {
+									const char = letterFor(g);
+									if (char !== undefined) {
+										parsedLookup["input"].push(char);
+									}
+								}
+							}
```
Comment (pascalw) [inline src/fondue/Fondue.js:729]: This block seems to very similar to `glyphToLetter`. Is it possible to use that here?

File: src/fondue/Fondue.js [hunk 2]
```diff
@@ -760,7 +777,45 @@ export default class Fondue {
 				});
 			}
 
-			return currentAllGlyphs;
+			// Chained context substitution
+			if (lookup.lookupType === 6) {
```
Comment (pascalw) [inline src/fondue/Fondue.js:781]: Just a thought: It might be nice to split this big function up in several small functions, one per lookup type. That would make it a lot easier to find the implementation for a specific lookup type, instead of having to scroll through this large function to find the right section. WDYT?

File: src/fondue/Fondue.js [hunk 3]
```diff
@@ -760,7 +777,45 @@ export default class Fondue {
...
+				lookup.subtableOffsets.forEach((_, i) => {
+					let subtable = lookup.getSubTable(i);
```
Comment (pascalw) [inline src/fondue/Fondue.js:783]: `let` -> `const`?

File: src/fondue/Fondue.js [hunk 4]
```diff
@@ -760,7 +777,45 @@ export default class Fondue {
...
+
+					if (subtable.inputGlyphCount > 0) {
+						subtable.inputCoverageOffsets.forEach((offset) => {
+							const coverage = subtable.getCoverageFromOffset(
+								offset
+							);
+							parsedLookup["input"][i] = glyphToLetter(coverage);
```
Comment (pascalw) [inline src/fondue/Fondue.js:790]: Since you're assigning based on index (`i`) here and these assignments are wrapped in conditions, it's possible that `parsedLookup["input"]` contains 'gaps`, i.e. an array with `[undefined, 'A', undefined, undefined, 'B']` etc. Is that intentional and something that can actually happen? It feels to me like this might be quite surprising as a consumer of this function.

File: src/fondue/Fondue.js [hunk 5]
```diff
@@ -760,7 +777,45 @@ export default class Fondue {
...
+						});
+					}
+
+					if (subtable.backtrackGlyphCount > 0) {
+						subtable.backtrackCoverageOffsets.forEach((offset) => {
+							const coverage = subtable.getCoverageFromOffset(
```
Comment (pascalw) [inline src/fondue/Fondue.js:796]: `getCoverageFromOffset` is performed for input, backtrack and lookahead. I think this could be one only once? Not sure if it's an expensive operation or not.

File: src/fondue/Fondue.js [hunk 6]
```diff
@@ -779,22 +834,20 @@ export default class Fondue {
...
+						allGlyphs[script][lang][feature.featureTag].push(
+							parseLookup(lookup)
 						);
 					});
 				});
 			});
 		});
 
+		// console.log(allGlyphs);
+
```
Comment (pascalw) [inline src/fondue/Fondue.js:850]: You forgot a console.log here.

File: test/Fondue.test.js [hunk 7]
```diff
@@ -372,11 +391,15 @@ describe("Layout features", () => {
...
-				liga: {
-					alternateCount: [],
-					input: ["ffl", "ffi", "ff", "fl", "fi"],
-					type: 4,
-				},
+				liga: [
+					{
+						alternateCount: [],
+						lookahead: [],
+						backtrack: [],
```
Comment (pascalw) [inline test/Fondue.test.js:398]: Could you add a test where these values are not empty? I guess the code you've added is not really exercised in tests now, since all these tests have empty `lookahead` and `backtrack` values.

File: src/fondue/Fondue.js (hunk 1 above)
Comment (RoelN) [reply src/fondue/Fondue.js:729]: While I was figuring out how this would work I thought there were more differences. But looking at the code now I think I can pass the index (`j`) to glyphToLetter and have it work with that.

File: src/fondue/Fondue.js (hunk 2 above)
Comment (RoelN) [reply src/fondue/Fondue.js:781]: Maybe even export this to a separate file, as this is really going to be the biggest chunk of logic in the engine. The rest is basically massaging data to be easier consumable by the site, but this is a lot more complex. is there a structure I could follow here, like a best practise?

File: src/fondue/Fondue.js (hunk 4 above)
Comment (RoelN) [reply src/fondue/Fondue.js:803]: `glyphToLetter` will actually return an array of characters for an input/backtrack/lookahead.

These are empty by default (done [here](https://github.com/Wakamai-Fondue/wakamai-fondue-engine/blob/c3bd44a214157e42ae1183bd2144d6040fedd279/src/fondue/Fondue.js#L685)) and will only be filled when this condition is true. Either outcome is fine and expected.

So this is working differently than you think, I think :-) Does that make sense?

File: test/Fondue.test.js (hunk 7 above)
Comment (RoelN) [reply test/Fondue.test.js:398]: Will add, thanks! 👍

File: src/fondue/Fondue.js (hunk 2 above)
Comment (pascalw) [reply src/fondue/Fondue.js:794]: Indeed. What do you think about something like this?

```js
parseLookup(lookup) {
  const lookupParsers = {
    1: require("./lookups/1.js"),
    2: require("./lookups/1.js"),
    ...
  };
  
  return lookupParsers[lookup.lookupType](lookup);
}
```

Where every file in `lookups/` would export a function, that takes the lookup (or whatever arguments are needed).
I hope you can come up with better names than `1.js`, but you get the point :-)

File: src/fondue/Fondue.js (hunk 4 above)
Comment (pascalw) [reply src/fondue/Fondue.js:803]: Right, I see. `glyphToLetter` is perhaps a bit misleading name then?

File: src/fondue/Fondue.js (hunk 1 above)
Comment (RoelN) [reply src/fondue/Fondue.js]: Updated! Much cleaner, thanks for the pointer!

File: src/fondue/Fondue.js (hunk 5 above)
Comment (RoelN) [reply src/fondue/Fondue.js]: Wowzers!! I was staring at this code for so long, I didn't see the obvious. Great simplification, thanks!

File: src/fondue/Fondue.js (hunk 4 above)
Comment (RoelN) [reply src/fondue/Fondue.js]: Could you take a look at the refactored code to see if it's still confusing?

File: src/fondue/Fondue.js (hunk 2 above)
Comment (RoelN) [reply src/fondue/Fondue.js:793]: Thanks! That sounds really good. I think I can come up with better names :-)

To not further delay this particular PR, I've moved this to an issue so we can keep track of it: https://github.com/Wakamai-Fondue/wakamai-fondue-engine/issues/44

File: src/fondue/Fondue.js (hunk 4 above)
Comment (pascalw) [reply src/fondue/Fondue.js]: Makes sense now! :+1:

File: test/Fondue.test.js (hunk 7 above)
Comment (RoelN) [reply test/Fondue.test.js]: Done!

Also added checks for the new `summary` thing.

File: src/fondue/Fondue.js [hunk 8]
```diff
@@ -649,60 +649,93 @@ export default class Fondue {
...
+		function mergeUniqueCoverage(existing, addition) {
+			return [...new Set([...(existing || []), ...addition])];
+		}
+
+		// Returns glyphs that are mapped directly to characters for
+		// this coverage. If a glyph maps to another glyph, it's
+		// ignored.
+		// If only a specific rangeRecord needs to be processed, e.g.
+		// for lookup type 3, you can pass the desired index.
+		function charactersFromGlyphs(coverage, index) {
```
Comment (pascalw) [inline src/fondue/Fondue.js:680]: Could we extract all these inner functions into private functions on Fondue.js? Or perhaps extract the entire type 6 lookup to a separate file? This function is pretty big now (~ 300 lines) and it feels like there are a lot of details that are not really important to understanding the flow of the type 6 lookup. E.g. it's good to know reading this code that it's looking up chars from glyphs, but _how_ that's done is not really relevant when you try to just understand the type 6 lookup, right?

File: src/fondue/Fondue.js [hunk 9]
```diff
@@ -649,60 +649,93 @@ export default class Fondue {
...
+					.map(letterFor);
 			}
+			return results;
+		}
+
+		function parseLookup(lookup) {
+			const parsedLookup = {
+				type: lookup.lookupType,
+				typeName:
+					lookupTypes[lookup.lookupType] || "Unknown lookup type",
```
Comment (pascalw) [inline src/fondue/Fondue.js:718]: Shouldn't this function fail if an unknown lookup type is passed? It seems like this function is not doing much if you pass in an unsupported lookup type, so it might as well fail more clearly IMO. WDYT?

File: src/fondue/Fondue.js [hunk 10]
```diff
@@ -760,10 +794,112 @@ export default class Fondue {
...
+					let inputChars;
+					let backtrackChars;
+					let lookaheadChars;
+
+					if (subtable.inputGlyphCount > 0) {
+						subtable.inputCoverageOffsets.forEach((offset) => {
+							const coverage = subtable.getCoverageFromOffset(
+								offset
+							);
+							inputChars = charactersFromGlyphs(coverage);
```
Comment (pascalw) [inline src/fondue/Fondue.js:828]: This lambda will get executed multiple times potentially, but it always overwrites the `inputChars` variable. Is that intentional? E.g. does only the last offset matter? Same for all lookups below.

File: src/fondue/Fondue.js [hunk 11]
```diff
@@ -649,60 +649,93 @@ export default class Fondue {
...
+		}
+
+		function parseLookup(lookup) {
+			const parsedLookup = {
+				type: lookup.lookupType,
+				typeName:
+					lookupTypes[lookup.lookupType] || "Unknown lookup type",
+				input: [],
+				backtrack: [],
+				lookahead: [],
```
Comment (pascalw) [inline src/fondue/Fondue.js:721]: It seems like these fields belong to only lookupType 6, but they are returned for any lookup type now. Perhaps you could declare this for each lookup type specifically?

File: src/fondue/Fondue.js [hunk 12]
```diff
@@ -798,3 +942,76 @@ export default class Fondue {
...
+		return array;
+	};
+
+	// Create some kind of "all backtracks" or "all lookaheads"
+	for (const lookup of feature["lookups"]) {
+		if (lookup.type !== 6) continue;
+
+		// Create all possible combinations of input, backtrack and lookahead
+		for (const key in Object.entries(lookup["input"])) {
+			allInputs = [...new Set(allInputs.concat(lookup["input"][key]))];
```
Comment (pascalw) [inline src/fondue/Fondue.js:967]: Given that you're creating Sets here again, it's probably going to be a lot more efficient if you change the whole lookup 6 type code to use Sets instead of Arrays. Certainly if the domain is such that these are actually sets and not arrays. WDYT?

File: src/fondue/Fondue.js [hunk 13]
```diff
@@ -798,3 +942,76 @@ export default class Fondue {
...
+
+	if (allLookaheads.length) {
+		allCombinations.push(allLookaheads.slice(0, limit));
+	}
+
+	let summarizedCombinations = allCombinations
+		.reduce((a, b) =>
+			a.reduce((r, v) => r.concat(b.map((w) => [].concat(v, w))), [])
+		)
+		.map((a) => a.join(""));
```
Comment (pascalw) [inline src/fondue/Fondue.js:1009]: It's not really clear to me what this does. Can you either clarify the code, or add some comments?

File: test/Fondue.test.js [hunk 14]
```diff
@@ -342,45 +346,116 @@ describe("Layout features", () => {
...
+				alternateCount: [],
 			})
 		);
 	});
 
+	test("returns lookup 6 layout features", async () => {
+		const fondue = await AbrilFatFaceFont();
+
+		expect(
+			// Yes, four empty spaces (artifact of old font editing software)
```
Comment (pascalw) [inline test/Fondue.test.js:401]: Ah my grandfather used to tell me these stories ;-)

File: test/Fondue.test.js [hunk 15]
```diff
@@ -342,45 +346,116 @@ describe("Layout features", () => {
...
+			fondue.featureChars["    "]["dflt"]["calt"]["lookups"][0]
+		).toEqual({
+			alternateCount: [],
+			backtrack: [
+				["j"],
+				["g"],
+				["g"],
+				undefined,
```
Comment (pascalw) [inline test/Fondue.test.js:409]: What does undefined mean in the this context?

File: test/Fondue.test.js [hunk 16]
```diff
@@ -342,45 +346,116 @@ describe("Layout features", () => {
...
+	test("returns lookup 6 layout features", async () => {
+		const fondue = await AbrilFatFaceFont();
+
+		expect(
+			// Yes, four empty spaces (artifact of old font editing software)
+			fondue.featureChars["    "]["dflt"]["calt"]["lookups"][0]
+		).toEqual({
+			alternateCount: [],
+			backtrack: [
+				["j"],
```
Comment (pascalw) [inline test/Fondue.test.js:406]: Can this return anything other than one character? If not, this doesn't have to be an array. If it can, this is missing some test coverage?

File: test/Fondue.test.js [hunk 17]
```diff
@@ -342,45 +346,116 @@ describe("Layout features", () => {
...
+
+	test("returns lookup 6 summary", async () => {
+		const fondue = await AbrilFatFaceFont();
+
+		expect(
+			// Yes, four empty spaces (artifact of old font editing software)
+			fondue.featureChars["    "]["dflt"]["calt"]["summary"][
+				"summarizedCombinations"
+			].length
+		).toBeGreaterThanOrEqual(1);
```
Comment (pascalw) [inline test/Fondue.test.js:456]: Can you make this test a bit more specific, e.g. testing the actual returned structure? This will help with not breaking backwards compatibility in the future. And provide some documentation.

**Comment by RoelN** [comment]
Thanks for the extensive reviews, @pascalw. Like discussed offline, I'll merge this now so we can go ahead with the beta, and either improve things in a new PR, or create issues to keep track of them!
